docker-compose-junit-jupiter/src/main/java/com/palantir/docker/compose/DockerComposeExtension.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.docker.compose;
+
+import static com.palantir.docker.compose.connection.waiting.ClusterHealthCheck.serviceHealthCheck;
+import static com.palantir.docker.compose.connection.waiting.ClusterHealthCheck.transformingHealthCheck;
+
+import com.palantir.docker.compose.configuration.DockerComposeFiles;
+import com.palantir.docker.compose.configuration.ShutdownStrategy;
+import com.palantir.docker.compose.connection.Container;
+import com.palantir.docker.compose.connection.DockerPort;
+import com.palantir.docker.compose.connection.waiting.ClusterHealthCheck;
+import com.palantir.docker.compose.connection.waiting.ClusterWait;
+import com.palantir.docker.compose.connection.waiting.HealthCheck;
+import com.palantir.docker.compose.logging.FileLogCollector;
+import java.util.List;
+import org.immutables.value.Value;
+import org.joda.time.ReadableDuration;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+@Value.Immutable
+@CustomImmutablesStyle
+public abstract class DockerComposeExtension extends DockerComposeManager
+        implements BeforeEachCallback, AfterEachCallback {
+
+    @Override
+    public void beforeEach(ExtensionContext context) throws Exception {
+        before();
+    }
+
+    @Override
+    public void afterEach(ExtensionContext context) throws Exception {
+        after();
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder extends ImmutableDockerComposeExtension.Builder implements BuilderExtensions {
```

Comment:
We should definitely just extend a single builder directly. You can get the builder to return a `DockerComposeManager` subclass like `DockerComposeExtension` by fiddling with one of the immtuables style options.

Reply:
I don't believe that's possible since the immutable classes do not have a common class heirarchy

Reply:
Yeah, it looks that way. My big worry is about all the duplication of common builder helper methods in 3 places - I'm wondering whether there is any way we can not repeat them using interface default methods or something similar.